pkg/cli/cluster_free_disk_space.go

```diff
@@ -0,0 +1,128 @@
+	v1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	rookcli "github.com/rook/rook/pkg/client/clientset/versioned"
+
+	clusterspace "github.com/replicatedhq/kurl/pkg/cluster/space"
+)
+
+const (
+	isDefaultStorageClassAnnotation = "storageclass.kubernetes.io/is-default-class"
+	openEBSLocalProvisioner         = "openebs.io/local"
+	rookRBDProvisioner              = "rook-ceph.rbd.csi.ceph.com"
+	rookCephFSProvisioner           = "rook-ceph.cephfs.csi.ceph.com"
+)
+
+// NewClusterCheckFreeDiskSpaceCmd returns a command that is capable of reporting back the amount of free space in the cluster
+// for a provided storage class.
+func NewClusterCheckFreeDiskSpaceCmd(cli CLI) *cobra.Command {
+	var forStorageClass, openEBSImage, openEBSNode string
+	logbuf := bytes.NewBuffer(nil)
+	logger := log.New(logbuf, "", 0)
+	k8sConfig := config.GetConfigOrDie()
+	clientSet := kubernetes.NewForConfigOrDie(k8sConfig)
+	rookClientSet := rookcli.NewForConfigOrDie(k8sConfig)
+
+	cmd := &cobra.Command{
+		Use:          "check-free-disk-space",
+		Short:        "Returns the amount of free disk space (in bytes) for a given storage class",
+		Example:      "kurl cluster check-free-disk-space --storageclass openebs",
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			classes, err := clientSet.StorageV1().StorageClasses().List(cmd.Context(), metav1.ListOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to list cluster storage classes: %w", err)
+			}
+
+			var selectedClass *v1.StorageClass
+			for _, class := range classes.Items {
+				val, ok := class.Annotations[isDefaultStorageClassAnnotation]
+				if ok && val == "true" && forStorageClass == "" {
+					selectedClass = &class
+					logger.Printf("Selected default storage class %q", selectedClass.Name)
+					break
+				}
+
+				if class.Name == forStorageClass {
+					selectedClass = &class
+					logger.Printf("Selected storage class %q", selectedClass.Name)
+					break
+				}
+			}
+
+			if selectedClass == nil {
+				if forStorageClass == "" {
+					return fmt.Errorf("failed to find default storage class")
+				}
+				return fmt.Errorf("failed to find storage class %q", forStorageClass)
+			}
+
+			switch selectedClass.Provisioner {
+			case openEBSLocalProvisioner:
+				freeSpaceGetter, err := clusterspace.NewOpenEBSFreeDiskSpaceGetter(clientSet, logger, openEBSImage, selectedClass.Name)
+				if err != nil {
+					fmt.Print(logbuf.String())
+					return fmt.Errorf("failed to start openebs free space getter: %w", err)
+				}
+
+				volumes, err := freeSpaceGetter.OpenEBSVolumes(cmd.Context())
+				if err != nil {
+					fmt.Print(logbuf.String())
+					return fmt.Errorf("failed to get openebs free space: %w", err)
+				}
+
+				for node, volume := range volumes {
+					if openEBSNode == "" {
+						fmt.Printf("%d\t%s\n", volume.Free, node)
```

Comment:
Question how is the expected result? 

value
value
value 
error? 

Should we not show the total? Could you please add an screenshot of your test for we see how that would to be?

Reply:
No, we must show per node (for openebs). There is an example in the description of this PR.

Reply:
@ricardomaraschini see that in this implementation we will

print value 
print value
error found and returned / stop

Example:

$ /tmp/kurl cluster free-disk-space --storageclass openebs
62421426176     ip-172-31-60-177
62894407680     ip-172-31-80-103
Error 

Therefore, I think we should change that for return the error only after check all values OR just return the output when errors are not found. WDYT is that make sense?

Reply:
This output won't happen. Or we print the output or the error. Why do you think it is possible ?

Reply:
HI @ricardomaraschini, 

> This output won't happen. Or we print the output or the error. Why do you think it is possible ?

We have a loop and for each time we print the value see:
> fmt.Printf("%d\t%s\n", volume.Free, node) 
Then, if an error occurs in the loop we return an error and we stop to check.

So, how the scenario above will not happens?
To do what you are saying here `Or we print the output or the error` we can not use fmt.Print in the loop. We need to store the values in array and just print them after all process.`

Reply:
I see here we will continue. Sorry, I did not look this one properly.